icloud.com: duplicate and strangely offset highlight in icloud.com notes

jesxilin opened the pull request:
#### 0bbbd26aee6010a09afb73db99aad5603a98a6fa
<pre>
icloud.com: duplicate and strangely offset highlight in icloud.com notes
<a href="https://bugs.webkit.org/show_bug.cgi?id=282102">https://bugs.webkit.org/show_bug.cgi?id=282102</a>
<a href="https://rdar.apple.com/137094713">rdar://137094713</a>

Reviewed by Wenson Hsieh.

icloud.com renders their own selection in their style so we should
hide our real selection UI. However, we display our selection as well
in addition to icloud.com notes which results in the strange duplicate offset
selection being shown.

We should change the logic of supressing selection to include
if selection container element is clipped/transparent instead of just if the
editable root element is clipped/transparent.
In this example, when we are not editing, the duplicate selection is shown.

Finally, added a LayoutTest to confirm this new behavior works as intended by selecting
something noneditable in a hidden container.

* LayoutTests/editing/selection/ios/hide-selection-in-hidden-container-noneditable-expected.txt: Added.
* LayoutTests/editing/selection/ios/hide-selection-in-hidden-container-noneditable.html: Added.
* Source/WebCore/dom/Position.h:
* Source/WebKit/Shared/EditorState.h:
* Source/WebKit/Shared/EditorState.serialization.in:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView _updateSelectionAssistantSuppressionState]):
(-[WKContentView hasHiddenContentEditable]):
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::updateCursorAccessoryPlacement):
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::isTransparentOrFullyClipped const):
(WebKit::WebPage::getPlatformEditorStateCommon const):
* Source/WebKit/WebProcess/WebPage/WebPage.h:

Canonical link: <a href="https://commits.webkit.org/285925@main">https://commits.webkit.org/285925@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/be4671a71264c8e052c551a6e3efde39516ccd62

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/74176 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/53605 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/26987 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/78549 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/25414 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/76293 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/62738 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/1390 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/58311 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/16653 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/77243 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/48475 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/63806 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/38721 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/45396 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/21302 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/23747 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/66859 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/21648 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/80069 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/1493 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/837 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/66614 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/1638 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/63823 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/65887 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16349 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/9829 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/7981 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/1457 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/4245 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/1486 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/1474 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/1493 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on LayoutTests/editing/selection/ios/hide-selection-in-hidden-container-noneditable.html:
```diff
@@ -0,0 +1,54 @@
+    hiddenDiv.append(pText);
+
+    console.log("4. Wait for selection to disappear.");
+    await UIHelper.waitForSelectionToDisappear();
```
Could we double check that this test fails without the fix in this patch?

Edit: we should add this to the end of the test:

```
shouldBe("getSelection().rangeCount", "1");
```

...to verify that the selection isn't being cleared only due to `pText` being reparented.

jesxilin replied:
Added the `shouldBe` to the test, but it passes without the fix... Is there a way to "debug" the test when running (e.g. breakpoints, printing)?

whsieh replied:
> Added the shouldBe to the test, but it passes without the fix.

Oh, my bad — reparenting the selection container _collapses_ the DOM selection rather than clearing it altogether, so the `rangeCount` will still be equal to 1. Can we instead check this?

```
shouldBeEqualToString("getSelection().type", "Range");
```

This should confirm that the contents of `pText` remain selected, after making it hidden.

> Is there a way to "debug" the test when running (e.g. breakpoints, printing)

Yep, both are viable approaches for debugging tests (e.g. `WTFLogAlways`, or using `lldb -n WebKitTestRunnerApp -w` or `lldb -n com.apple.WebKit.WebContent.Development -w`).

whsieh commented on Source/WebCore/dom/Position.h:
```diff
@@ -199,6 +199,8 @@ class Position {
     // FIXME: Combine this function with operator==
     bool equals(const Position&) const;
 
+    WEBCORE_EXPORT static Node* getCommonAncestorForPositions(const Position&, const Position&);
```
Perhaps we could just `WEBCORE_EXPORT` `commonInclusiveAncestor` instead of introducing a new method?

pxlcoder commented on Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
```diff
@@ -578,9 +578,9 @@
+bool WebPage::isTransparentOrFullyClipped(const Node& node) const
 {
-    auto* renderer = element.renderer();
+    auto* renderer = node.renderer();
```
```suggestion
    CheckedPtr renderer = node.renderer();
```

whsieh commented on Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
```diff
@@ -670,12 +670,19 @@
 #if PLATFORM(IOS_FAMILY)
         result.visualData->editableRootBounds = rootViewInteractionBounds(Ref { *editableRootOrFormControl });
 #endif
+    } else {
```
We should probably avoid doing this ancestor walk in the case where the selection is a caret (i.e. collapsed).
```suggestion
    } else if (result.selectionIsRange) {
```

whsieh commented on Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
```diff
@@ -670,12 +670,19 @@
+        auto posEnd = selection.end();
+
+        RefPtr ancestorContainer = Position::getCommonAncestorForPositions(posStart, posEnd);
+        postLayoutData.selectionIsTransparentOrFullyClipped = isTransparentOrFullyClipped(*ancestorContainer);
```
Probably best to be robust in the case where there's no common ancestor (instead of crashing)

```suggestion
        if (RefPtr ancestorContainer = commonInclusiveAncestor(selection.start(), selection.end()))
            postLayoutData.selectionIsTransparentOrFullyClipped = isTransparentOrFullyClipped(*ancestorContainer);
```

pxlcoder replied:
Should `ancestorContainer` be null-checked?

whsieh commented on Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
```diff
@@ -670,12 +670,16 @@
+            postLayoutData.selectionIsTransparentOrFullyClipped = isTransparentOrFullyClipped(*ancestorContainer);
     }
 
+
```
Nit - extra newline.

whsieh commented on Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
```diff
@@ -670,12 +670,16 @@
         result.visualData->editableRootBounds = rootViewInteractionBounds(Ref { *editableRootOrFormControl });
 #endif
+    } else if (result.selectionIsRange) {
+        if (RefPtr ancestorContainer = WebCore::commonInclusiveAncestor(selection.start(), selection.end()))
```
```suggestion
        if (RefPtr ancestorContainer = commonInclusiveAncestor(selection.start(), selection.end()))
```

I think you can omit `WebCore::` here.

jesxilin committed:
icloud.com: duplicate and strangely offset highlight in icloud.com notes
https://bugs.webkit.org/show_bug.cgi?id=282102
rdar://137094713

Reviewed by Wenson Hsieh.

icloud.com renders their own selection in their style so we should
hide our real selection UI. However, we display our selection as well
in addition to icloud.com notes which results in the strange duplicate offset
selection being shown.

We should change the logic of supressing selection to include
if selection container element is clipped/transparent instead of just if the
editable root element is clipped/transparent.
In this example, when we are not editing, the duplicate selection is shown.

Finally, added a LayoutTest to confirm this new behavior works as intended by selecting
something noneditable in a hidden container.

* LayoutTests/editing/selection/ios/hide-selection-in-hidden-container-noneditable-expected.txt: Added.
* LayoutTests/editing/selection/ios/hide-selection-in-hidden-container-noneditable.html: Added.
* Source/WebCore/dom/Position.h:
* Source/WebKit/Shared/EditorState.h:
* Source/WebKit/Shared/EditorState.serialization.in:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView _updateSelectionAssistantSuppressionState]):
(-[WKContentView hasHiddenContentEditable]):
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::updateCursorAccessoryPlacement):
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::isTransparentOrFullyClipped const):
(WebKit::WebPage::getPlatformEditorStateCommon const):
* Source/WebKit/WebProcess/WebPage/WebPage.h:

Canonical link: https://commits.webkit.org/285925@main
